net/tcp/tcp_sendfile.c

```diff
@@ -264,15 +274,43 @@ static uint16_t sendfile_eventhandler(FAR struct net_driver_s *dev,
           goto end_wait;
         }
 
-      /* No. Fall through to send more data if necessary */
+      /* Fast Retransmit (RFC 5681): an acknowledgment is considered a
+       * "duplicate" when (a) the receiver of the ACK has outstanding data,
+       * (b) the incoming acknowledgment carries no data, (c) the SYN and
+       * FIN bits are both off, (d) the acknowledgment number is equal to
+       * the greatest acknowledgment received on the given connection
+       * and (e) the advertised window in the incoming acknowledgment equals
+       * the advertised window in the last incoming acknowledgment.
+       */
+
+      if (pstate->snd_acked < pstate->snd_sent &&
+          (flags & TCP_NEWDATA) == 0 &&
+          (tcp->flags & (TCP_SYN | TCP_FIN)) == 0 &&
+          ackno == pstate->snd_prev_ack &&
+          conn->snd_wnd == pstate->snd_prev_wnd)
+        {
+          if (++pstate->snd_dup_acks >=
+                CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK)
+            {
+              flags |= TCP_REXMIT;
+              pstate->snd_dup_acks = 0;
```

Comment:
So in this case we reset fast retransmit logic right?

Reply:
I'm asking because logic becomes slightly different from original logic in `net/tcp/tcp_send_unbuffered.c`. I mean that before it was `N -> N -> N ->T -> T -> T -> T -> T -> T ...` and now it becomes `N -> N -> N ->T -> N -> N -> N -> T -> N ...`

Reply:
Yes. Concerning resetting it or not right after the condition (3 dup ACKs) is met, it seems it is not regulated by RFC 5681. It seems both algorithms are correct (let it reset only if the the main complex condition get false, or reset it after each 3 dup ACKs if the bad condition (massive dup ACKs) is going on).

Reply:
E.g. NuttX tcp_send_buffered and Ethernut implementations both reset the counter right after the counter gets 3 not waiting for the main complex condition getting false.
I have tested both algorithms on practice (on the real traffic with lost packets), the both algorithm versions seem to work well.